### PR TITLE
WIP: Provide bitMap at loading to avoid later calculations

### PR DESCRIPTION
**Pull request by Sceat**
### User story
While working on [@AresRPG/aresrpg-world](https://github.com/AresRPG/aresrpg-world) i was thinking about the best way to compute the bitmap before sending chunk, and in fact i think that it should be inside the first block reading iteration so we only iterate once.

### How
While testing some feature i needed a fast way to spawn in the world so i'm using a basic generation which provide bitMaps the same way it provide chunks, really usefull and a litle perf gains imo
> here there is only one type of chunks so the `x, z` value are obsolete but it was more for the form and futur implementations
```js
import { Chunk, Block } from '@/constant'
import { Vec3 } from 'vec3'

export default class {
	mask = 0

	constructor(chunkPredicate) {
		this.chunkPredicate = chunkPredicate
	}

	updateBitMap(y) {
		y = Math.floor(y / 16)
		this.mask |= 1 << y
	}

	bitMap(x, z) {
		return this.mask
	}

	load(x, z) {
		if (!this.chunkPredicate(x, z)) return
		const chunk = new Chunk()
		chunk.initialize((x, y, z) => {
			chunk.setSkyLight(new Vec3(x, y, z), 15)
			if (y === 10) {
				this.updateBitMap(y)
				return new Block(9, 3)
			}
			return new Block(0)
		})
		return chunk
	}
}
```
---
### WIP
This PR isn't tested nor finished, i create it so everyone can say what they think

current state below, no ETA
```js
  // returns a Promise. Resolve a Chunk object or reject if it hasn’t been generated
  async load (x, z) {
    const data = await this.loadRaw(x, z)
    if (data == null) { return null }
    const { chunk, bitMap } = nbtChunkToPrismarineChunk(data) // TODO: think about the best way to export the bitMap
    return chunk
  }
```
The idea is basic, i followed the current code structure and added a object containing the mask for the current readed column, and swipe a bit if the blockType is not `AIR`

```js
function readBlocks (chunk, mask, sectionY, blocks) {
  blocks = Buffer.from(blocks)
  for (let index = 0; index < blocks.length; index++) {
    const blockType = blocks.readUInt8(index)
    const pos = indexToPos(index, sectionY)
    if(blockType) updateBitMap(mask, pos.y) // if not "empty"
    chunk.setBlockType(pos, blockType)
  }
}
```

### CODE REVIEW

**Comment by Sceat** [comment]
Just need to test it and it will be done

**Review by Sceat — COMMENTED** [review]
seems to be the best options, what you guys think ?

File: src/anvil.js [hunk 1]
```diff
@@ -34,11 +34,11 @@ class Anvil {
   }
 
   // returns a Promise. Resolve a Chunk object or reject if it hasn’t been generated
-  async load (x, z) {
+  async load (x, z, withBitMap = false) {
```
Comment (Sceat) [inline src/anvil.js:37]: ^ 
_(i missclicked review instead of comment)_

**Comment by Sceat** [comment]
likely ready but depends on [@protocol/#608](https://github.com/PrismarineJS/node-minecraft-protocol/issues/608) to be tested in a healthy way

File: src/chunk.js [hunk 2]
```diff
@@ -73,15 +74,21 @@ function indexToPos (index, sectionY) {
...
     const pos = indexToPos(index, sectionY)
+    if(blockType) updateBitMap(mask, pos.y) // if not "empty"
     chunk.setBlockType(pos, blockType)
   }
 }
 
+function updateBitMap(mask, y) {
+  y = Math.floor(y / 16)
+  mask.bitMap |= 1 << y
+}
```
Comment (Sceat) [inline src/chunk.js:90]: **TL;DR** everything is here ^

image

**Comment by rom1504** [comment]
Maybe this could be useful (cc @Karang since you're looking at this lib)

**Comment by Karang** [comment]
> Maybe this could be useful (cc @Karang since you're looking at this lib)

prismarine-chunk already contains the sectionMask https://github.com/PrismarineJS/prismarine-chunk/blob/master/src/pc/1.9/ChunkColumn.js#L12

Maybe at the time this PR was created it wasnt the case, but I think this can be closed now

**Comment by Sceat** [comment]
> but I think this can be closed now

Indeed, i'm totally under external load anyway 😕

**Comment by rom1504** [comment]
Ah right I misread this, I thought it was doing something to get the bitmask correctly defined in prismarine-chunk

**Pull request closed without merging**